### PR TITLE
Making object pointer copy-prop (slightly) more agressive in loop prepass

### PR DESCRIPTION
**Pull request by rajatd**
Object pointer copy prop in pre-pass will only kick in if both the object sym and its copy prop sym are single-def. I'm changing it such that the we'll do the optimization in prepass if - 

both object sym and copy sym should not be live on back-edge, if they are then they should not be assigned to inside the current loop or any of its parent loops.

### CODE REVIEW

File: lib/Backend/FlowGraph.cpp [hunk 1]
```diff
@@ -3645,6 +3645,26 @@ Loop::SetLoopTopInstr(IR::LabelInstr * loopTop)
     this->loopTopLabel = loopTop;
 }
 
+bool
+Loop::IsSymAssignedToInSelfOrParents(StackSym * const sym) const
+{
+    if (!this->symsAssignedToInLoop->Test(sym->m_id))
+    {
+        for (Loop* curLoop = this->parent; curLoop != nullptr; curLoop = curLoop->parent)
```
Comment (sigatrev) [inline lib/Backend/FlowGraph.cpp]: Might be a little cleaner to just initialize the loop to `this`

File: lib/Backend/GlobOpt.cpp [hunk 2]
```diff
@@ -5488,6 +5489,21 @@ GlobOpt::IsPrepassSrcValueInfoPrecise(IR::Opnd *const src, Value *const srcValue
...
+bool
+GlobOpt::IsSafeToTransferInPrepass(StackSym * const srcSym, ValueInfo *const srcValueInfo) const
+{
+    int32 intConstantValue;
+    return
+        srcSym->IsFromByteCodeConstantTable() ||
+        (
+            srcValueInfo->TryGetIntConstantValue(&intConstantValue) &&
+            !Js::TaggedInt::IsOverflow(intConstantValue) &&
+            GetTaggedIntConstantStackSym(intConstantValue) == srcSym
```
Comment (LouisLaf) [inline lib/Backend/GlobOpt.cpp]: I see the same code above...  What is it about?

Comment (rajatd) [reply lib/Backend/GlobOpt.cpp]: I will be reusing this code in the above function in a subsequent PR. Just wanted to change the object-pointer-copy-prop semantics in this PR.

---
In reply to: [180214048](https://github.com/Microsoft/ChakraCore/pull/4943#discussion_r180214048) [](ancestors = 180214048)

File: lib/Backend/FlowGraph.cpp (hunk 1 above)
Comment (rajatd) [reply lib/Backend/FlowGraph.cpp]: Yep, changed.

---
In reply to: [180184073](https://github.com/Microsoft/ChakraCore/pull/4943#discussion_r180184073) [](ancestors = 180184073)

**Comment by rajatd** [comment]
@dotnet-bot test OSX static_osx_osx_test

**Pull request merged**